### PR TITLE
Tree-like Rewriter

### PR DESCRIPTION
**Pull request by marcandre**
This is only a proof of concept for discussion for #401. If there is interest, it could become a real PR with a preference switch to maintain backward compatibility.

I've kept the 21 failing tests to make it easier to judge the differences. I'm still quite curious as to what kind of use case required these.

Here's the list:
```

TestSourceRewriter#test_insertion_within_replace_clobber [test_source_rewriter.rb:182]:
Parser::ClobberingError expected but nothing was raised.

TestSourceRewriter#test_remove_overlapping_insertion_clobber [test_source_rewriter.rb:212]:
Parser::ClobberingError expected but nothing was raised.

TestSourceRewriter#test_overlapping_delete:
Parser::ClobberingError: Trying to combine overlapping ranges: ...
test_source_rewriter.rb:458:in `test_overlapping_delete'

TestSourceRewriter#test_replace_overlapping_insertion_clobber [test_source_rewriter.rb:202]:
Parser::ClobberingError expected but nothing was raised.

TestSourceRewriter#test_overlapping_replace_left_smaller_than_replaced_matching:
Parser::ClobberingError: Trying to combine overlapping ranges: 
...test_source_rewriter.rb:312:in `test_overlapping_replace_left_smaller_than_replaced_matching'

TestSourceRewriter#test_overlapping_replace_left_larger_right_larger_matching:
Parser::ClobberingError: Trying to combine overlapping ranges: 
...test_source_rewriter.rb:338:in `test_overlapping_replace_left_larger_right_larger_matching'

TestSourceRewriter#test_remove_overlapping_multi_insertion_clobber [test_source_rewriter.rb:252]:
Parser::ClobberingError expected but nothing was raised.

TestSourceRewriter#test_subsuming_replace_both_smaller_non_matching [test_source_rewriter.rb:373]:
Parser::ClobberingError expected but nothing was raised.

TestSourceRewriter#test_clobbering_error_backward_compatibility [test_source_rewriter.rb:449]:
Expected false to be truthy.

TestSourceRewriter#test_replace_overlapping_multi_insertion_clobber [test_source_rewriter.rb:242]:
Parser::ClobberingError expected but nothing was raised.

TestSourceRewriter#test_overlapping_replace:
Parser::ClobberingError: Trying to combine overlapping ranges: 
...test_source_rewriter.rb:466:in `test_overlapping_replace'

TestSourceRewriter#test_insertion_within_remove_clobber [test_source_rewriter.rb:192]:
Parser::ClobberingError expected but nothing was raised.

TestSourceRewriter#test_subsuming_replace_both_larger_non_matching [test_source_rewriter.rb:391]:
Parser::ClobberingError expected but nothing was raised.

TestSourceRewriter#test_insertion_on_merged_insertion_clobber [test_source_rewriter.rb:266]:
Parser::ClobberingError expected but nothing was raised.

TestSourceRewriter#test_transaction_rollback [test_source_rewriter.rb:533]:
Parser::ClobberingError expected but nothing was raised.

TestSourceRewriter#test_replace_same_begin_larger_than_replaced_range_non_matching [test_source_rewriter.rb:301]:
Parser::ClobberingError expected but nothing was raised.

TestSourceRewriter#test_multiple_insertions_at_same_location_clobber [test_source_rewriter.rb:130]:
Parser::ClobberingError expected but nothing was raised.

TestSourceRewriter#test_clobber [test_source_rewriter.rb:415]:
Parser::ClobberingError expected but nothing was raised.

TestSourceRewriter#test_multi_insertion_within_replace_clobber [test_source_rewriter.rb:222]:
Parser::ClobberingError expected but nothing was raised.

TestSourceRewriter#test_multi_insertion_within_remove_clobber [test_source_rewriter.rb:232]:
Parser::ClobberingError expected but nothing was raised.

TestSourceRewriter#test_overlapping_replace_left_larger_right_smaller_matching:
Parser::ClobberingError: Trying to combine overlapping ranges: (rewriter):1:2 and (rewriter):1:4
test_source_rewriter.rb:330:in `test_overlapping_replace_left_larger_right_smaller_matching'
```

### CODE REVIEW

**Comment by marcandre** [comment]
I have a much improved version now. 
I added a new `TreeRewriter` class. It has really clear semantics and supports customizable policies with regards to certain scenarios like multiple replacements.
`Rewriter` uses `TreeRewriter`, tweaks some methods and uses a private policy so as to achieve a really good compatibility in most aspects with the existing version. In particular, the existing Rubocop test suite passes.
Some work remains with diagnostics, tests and I hope to find the time to write a more detailed report soon :smile:

**Comment by whitequark** [comment]
> Rewriter uses TreeRewriter

If you add a new Rewriter anyway, why not just deprecate the existing one but keep it?

**Comment by marcandre** [comment]
> If you add a new Rewriter anyway, why not just deprecate the existing one but keep it?

Right, it's more an exercise to be able to check `rubocop`'s test suite better against the new `TreeRewriter` than anything. I agree, for 100% compatibility, we can keep the existing `Rewriter` identical and deprecate it.

Actually, I see the following possibilities:

A) Keep `Parser::Source::Rewriter` and `Parser::Rewriter` intact, deprecate them and add `Parser::Source::TreeRewriter` and `Parser::TreeRewriter`.

B) Rename `Rewriter`s to `LegacyRewriter`s, and add the new `Rewriter`s. 

C) Replace `Rewriter` but have default policy matching reasonably closely the current `Rewriter`.

Any other suggestion?

I'm fine with any of these 3. For B & C, maybe wait for v2.5.0.0, along with dropping Ruby 1.9 compatibility?

It's quite difficult to predict incompatibilities in general with (B) and (C). They would probably be mostly about the `inserts`.
`Rubocop` is kind of an extreme example, because it silently swallows `ClobberingError`, and 24 specs would fail otherwise. Interestingly, 2 of those don't generate an error with the new `Rewriter` and pass.
For inserts, only one endpoint is considered currently, while my proposal always considers the range as a whole. It's impossible to know how (in)accurate the other end points could be, or what the effect would be. In `rubocop`'s case, there's one case where the `insert_before` have clearly wrong `end_pos` that are outside the document! (1 cop affecting 9 specs). Only one other spec produces a new `ClobberingError` because of taking into account the end point.

**Comment by whitequark** [comment]
Note that parser does *not* follow semver, so waiting for 2.5 doesn't make any sense. (A) is best.

**Comment by marcandre** [comment]
> Note that parser does not follow semver, so waiting for 2.5 doesn't make any sense.

Yes, I understand that. Ruby 2.5 is out in a month, so I thought hey, why not :smile:

Ok, so I'll tweak the PR for A).

I'll try to write up a short "upgrade path" in the doc.

**Comment by marcandre** [comment]
PS: Might as well revert my #7d679d7 commit, in case it creates any kind issue for some odd corner case. Do you want a PR for that?

**Comment by whitequark** [comment]
Reverted

**Comment by alexdowad** [comment]
If Rubocop has a cop which tries to modify code outside the boundaries of the document, that should be fixed. Why not submit a patch to Rubocop? After it is merged, Parser can be fixed without breaking Rubocop.

**Comment by marcandre** [comment]
@alexdowad I [filed an issue](https://github.com/bbatsov/rubocop/issues/5084) with Rubocop. Even then there would be at least another failure, and that's with a restrictive default policy which isn't the best.

**Comment by marcandre** [comment]
@whitequark I imagine #406 will land shortly, so it's ok if my code is Ruby 2+, right?

**Comment by whitequark** [comment]
Sure.

**Comment by marcandre** [comment]
Pushed a new version I'm really happy with. Should be ready for an actual code review.

Only thing I haven't changed: `bin/ruby-rewrite` still uses `Rewriter`. Should I change it to use `TreeRewriter`, or add CLI options for this?

**Comment by alexdowad** [comment]
I'm sure @whitequark will say not to change the existing behavior, but personally I would like to see `TreeRewriter` become the default eventually.

**Comment by marcandre** [comment]
Thanks @alexdowad , I appreciate that :smile: 
I think `TreeRewriter` will already be the new default, just in a non breaking way.
As for `bin/ruby-rewrite` I have no idea how much it's used, and if it's considered part of `parser`'s API.

**Comment by iliabylich** [comment]
How about `bin/ruby-tree-rewrite`?

**Comment by marcandre** [comment]
> How about bin/ruby-tree-rewrite?

Might be best to keep a single executable. I'd say either:
1) it uses TreeRewriter, or
2) uses TreeRewriter by default and has a switch, say `-x` `--legacy` for old Rewriter, or
3) uses Rewriter by default and has a switch, say `-t` `--tree-rewriter` for TreeRewriter.
Wouldn't be too hard (but slower) to
5) only switch given, use both to do the transform, compares the result. Warns if it gets different results.

**Comment by whitequark** [comment]
I don't understand. ruby-rewrite makes no assumptions about the class that a rewriter inherits from. It just looks up a constant based on the filename, you're free to use any of the basic rewriters.

**Comment by marcandre** [comment]
You may not have had the time to look at 8d95525ed515b2aaac2 where I change the API for `::Parser::Rewriter#rewrite`. I use keyword arguments and make the AST optional. I use this new API vs the old one to decide which of `Source::Rewriter` or `Source::TreeRewriter` to instantiate:

* rewrite(buffer, ast)  #=> Rewriter
* rewrite(buffer, ast: ast) or rewrite(buffer) # => TreeRewriter

It's also possible to pass the class to instantiate. In any case, `ruby-rewrite` calls `rewrite` on the given classes, so must decide which implementation to use.

**Comment by marcandre** [comment]
So I went ahead and implemented my option "2+3" for `ruby-rewrite`. In practice, for the vast majority of users, `ruby-rewrite` will produce the same output as before, while double checking that it is forward compatible with the `TreeRewriter`. In the rare case it isn't, a warning is issued and the legacy output is used. It's also possible to specify explicitly which rewriter to use, although I doubt this will be used much in practice.

Minor tweaks:
- I added `insert_{before|after}_multi` to the `TreeRewriter` (with deprecation warnings), in case anyone is using `@source_rewriter.insert_before_multi` and `ruby-rewrite`. I'm actually using this to test my warnings, as it's the only case I could come up with that would produce different but legit Ruby code.
- Improved the diagnostics.

I doubled-checked that `rubocop` passes with no errors on this branch, and has a minimal number of errors when using the `TreeRewriter` instead of the `Rewriter`.

On my end, this PR is ready to be merged.

**Comment by whitequark** [comment]
NAK on this magical autodetection. How about just making the interface identical for both rewriters? I don't see the inherent problem with this.

**Comment by marcandre** [comment]
There's `Rewriter` and `Source::Rewriter`. I'm not sure which you are referring to. I'm also not sure of which magical interface you are referring to.

Summary of the PR:
In my proposal `Source::Rewriter` is kept untouched, `Source::TreeRewriter` is added and has the exact same interface as `Source::Rewriter`.

Current `Rewriter` calls `Source::Rewriter.new`. I'm proposing to keep a single `Rewriter` class and be able to specify which of `Source::Rewriter` or `Source::TreeRewriter` to use (new API using keyword arguments — keyword `rewriter_class`). The old API, using sequential arguments, is still available but deprecated.

`bin/ruby-rewrite` should be able to use `Source::Rewriter` or `Source::TreeRewriter`. I'm proposing to add CLI options to do that (`--rewriter`). For the default behavior, I propose to keep compatibility using `Source::Rewriter` and warn if, and only if there would be a difference using the `Source::TreeRewriter`.

**Comment by marcandre** [comment]
@whitequark A friendly ping that I'm still waiting for guidance on the API changes, as summarized above.

**Comment by whitequark** [comment]
A very good way to have this merged quickly is a solution that has none of the following:
  * two executables,
  * CLI options,
  * rewriting everything twice and checking.

Right now, I do not fully understand how your proposal works and why it cannot be done without all of the above, and given that I do not have that much time available for Parser anymore, it's not going to change quickly.

**Comment by marcandre** [comment]
Thanks for the feedback.

I've amended my proposal: there's now `Parser::TreeRewriter` and `Parser::Rewriter` is deprecated. Both have the same API.

`bin/ruby-rewrite` no longer needs any change.

What do you think about this?

File: lib/parser/source/rewriter.rb [hunk 1]
```diff
@@ -2,17 +2,20 @@ module Parser
   module Source
 
     ##
-    # {Rewriter} performs the heavy lifting in the source rewriting process.
-    # It schedules code updates to be performed in the correct order and
-    # verifies that no two updates _clobber_ each other, that is, attempt to
-    # modify the same section of code. (However, if two updates modify the
-    # same section in exactly the same way, they are merged.)
+    # {Rewriter} is deprecated. Use {TreeRewriter} instead.
```
[the remark this answers is not in the document]
Comment (marcandre) [reply lib/parser/source/rewriter.rb]: Oh, right. Changed to use it for the deprecated classes.

File: lib/parser/source/tree_rewriter.rb [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+      end
+
+      def enforce_policy(event)
+        return if @policy[event] == :accept
+        return unless (values = yield)
+        trigger_policy(event, values)
+      end
+
+      POLICY_TO_LEVEL = {warn: :warning, raise: :error}.freeze
+      def trigger_policy(event, range: raise, conflict: nil, **arguments)
```
[the remark this answers is not in the document]
Comment (marcandre) [reply lib/parser/source/tree_rewriter.rb]: Yes. It's a private method though, I can remove it if you prefer.

**Comment by marcandre** [comment]
I guess the last question would be: should the deprecated API issue a warning (as in the PR), or should it be only deprecated in the doc for now, with a warning in a later release?

**Comment by whitequark** [comment]
A warning is fine, I think!

**Comment by marcandre** [comment]
Ok. Rebased. (and sorry, didn't realize it would un-approve the PR!)

**Comment by marcandre** [comment]
:tada:  Thank you @whitequark, @alexdowad & al. 😄

**Comment by whitequark** [comment]
Thanks for the great PR.

**Comment by marcandre** [comment]
Looks like version 2.4.1.0 is taking a while. Could we have a release 2.4.0.3 in the meantime?

**Comment by whitequark** [comment]
Unfortunately not, that's just as much work as doing 2.4.1.0.

**Pull request merged**